### PR TITLE
[SPARK-30937][DOC] Group Hive upgrade guides together

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This PR groups all hive upgrade related migration guides inside Spark 3.0 together.

Also add another behavior change of `ScriptTransform` in the new Hive section.

### Why are the changes needed?

Make the doc more clearly to user.

### Does this PR introduce any user-facing change?

No, new doc for Spark 3.0.

### How was this patch tested?

N/A.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan @gatorsmile

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -335,6 +335,10 @@ license: |
   
   - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3. You need to migrate your custom SerDes to Hive 2.3 or build your own Spark with `hive-1.2` profile. See HIVE-15167 for more details.
 
+  - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3. The result decimal represent in string can be different between Hive 1.2 and Hive 2.3 when use `TRANSFORM` syntax in sql. In Hive 1.2, decimal will eliminate all non-significant zeros in precision when represents in string. But in Hive 2.3, it will always show 18 digits in precision and make up with zero if its precision is less than 18.
```
Comment (srowen) [inline docs/sql-migration-guide.md]: "The string representation of a decimal ..."
use -> using
sql -> SQL
"In Hive 1.2, the string representation omits trailing zeroes. But in Hive 2.3, it is always padded to 18 digits with trailing zeroes if necessary"

File: docs/sql-migration-guide.md [hunk 2]
```diff
@@ -335,6 +335,10 @@ license: |
...
+  
+  - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3. The local file system has changed from `org.apache.hadoop.fs.LocalFileSystem` to `org.apache.hadoop.hive.ql.io.ProxyLocalFileSystem` when use Hadoop API, e.g. `new Path("file:/").getFileSystem(conf)`, and `fs.rename(Path src, Path dst)` will returns `true` in Hive 1.2 but `false` in Hive 2.3 if `dst` represents a file.
```
Comment (srowen) [inline docs/sql-migration-guide.md]: use -> using
API -> APIs
returns -> return

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/sql-migration-guide.md (hunk 1 above)
Comment (yhuai) [reply docs/sql-migration-guide.md]: Let's call it out transform is used for script transformation, which depends on hive's behavior. 

I was not clear how a hive change can affect spark. But they realized that transform is script transformation.

File: docs/sql-migration-guide.md (hunk 2 above)
Comment (yhuai) [reply docs/sql-migration-guide.md]: Should we group all of hive related behavior change under the same section or the same bullet point? Then, we make individual changes sub-bullet-points.

Also, when we say rename returns true for hive 1.2 when des exists, we should call out that the destination will get overwritten, right? Also, can we link to related hadoop and hive tickets?

Comment (Ngone51) [reply docs/sql-migration-guide.md]: > Should we group all of hive related behavior change under the same section or the same bullet point?

I'll make it in a separate PR.

**Comment by Ngone51** [comment]
Thanks for the review, updated. @srowen @yhuai

File: docs/sql-migration-guide.md [hunk 3]
```diff
@@ -335,6 +335,10 @@ license: |
   
   - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3. You need to migrate your custom SerDes to Hive 2.3 or build your own Spark with `hive-1.2` profile. See HIVE-15167 for more details.
 
+  - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3. The string representation of a decimal can be different between Hive 1.2 and Hive 2.3 when using `TRANSFORM` syntax in SQL for script transformation, which depends on hive's behavior. In Hive 1.2, the string representation omits trailing zeroes. But in Hive 2.3, it is always padded to 18 digits with trailing zeroes if necessary.
+  
+  - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3. The local file system has changed from `org.apache.hadoop.fs.LocalFileSystem` to `org.apache.hadoop.hive.ql.io.ProxyLocalFileSystem` when using Hadoop APIs, e.g. `new Path("file:/").getFileSystem(conf)`, and `fs.rename(Path src, Path dst)` will return `true` and overwrite `dst` in Hive 1.2 but return `false` in Hive 2.3 if `dst` is an existed file. See HIVE-13705 and HDFS-10385 for more details.
```
Comment (HyukjinKwon) [inline docs/sql-migration-guide.md]: How does this relate to Spark's behaviour?

Comment (cloud-fan) [reply docs/sql-migration-guide.md]: Yea, we should highlight the changes for end-users, not internal differences. If we can't find one, let's remove this item.

Comment (Ngone51) [reply docs/sql-migration-guide.md]: Removed.

**Comment by Ngone51** [comment]
retest this please.

File: docs/sql-migration-guide.md [hunk 4]
```diff
@@ -252,7 +252,7 @@ license: |
         </tr>
     </table>
     
-  - Since Spark 3.0, CREATE TABLE without a specific provider will use the value of `spark.sql.sources.default` as its provider. In Spark version 2.4 and earlier, it was hive. To restore the behavior before Spark 3.0, you can set `spark.sql.legacy.createHiveTableByDefault.enabled` to `true`.
+  - Since Spark 3.0, `CREATE TABLE` without a specific provider will use the value of `spark.sql.sources.default` as its provider. In Spark version 2.4 and earlier, it was hive. To restore the behavior before Spark 3.0, you can set `spark.sql.legacy.createHiveTableByDefault.enabled` to `true`.
```
Comment (Ngone51) [inline docs/sql-migration-guide.md]: minor fix: CREATE TABLE -> 'CREATE TABLE'

File: docs/sql-migration-guide.md [hunk 5]
```diff
@@ -330,10 +330,14 @@ license: |
 
   - Since Spark 3.0, `SHOW CREATE TABLE` will always return Spark DDL, even when the given table is a Hive serde table. For generating Hive DDL, please use `SHOW CREATE TABLE AS SERDE` command instead.
 
-  - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3. This may need to set `spark.sql.hive.metastore.version` and `spark.sql.hive.metastore.jars` according to the version of the Hive metastore.
+  - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3 and it brings following impacts:
+  
+    - This may need to set `spark.sql.hive.metastore.version` and `spark.sql.hive.metastore.jars` according to the version of the Hive metastore.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: `This may need` -> `You may need`

Comment (cloud-fan) [reply docs/sql-migration-guide.md]: `... of the Hive metastore you want to connect to`

File: docs/sql-migration-guide.md [hunk 6]
```diff
@@ -330,10 +330,14 @@ license: |
...
   For example: set `spark.sql.hive.metastore.version` to `1.2.1` and `spark.sql.hive.metastore.jars` to `maven` if your Hive metastore version is 1.2.1.
   
-  - Since Spark 3.0, we upgraded the built-in Hive from 1.2 to 2.3. You need to migrate your custom SerDes to Hive 2.3 or build your own Spark with `hive-1.2` profile. See HIVE-15167 for more details.
+    - You need to migrate your custom SerDes to Hive 2.3 or build your own Spark with `hive-1.2` profile. See HIVE-15167 for more details.
+
+    - The string representation of a decimal can be different between Hive 1.2 and Hive 2.3 when using `TRANSFORM` syntax in SQL for script transformation, which depends on hive's behavior. In Hive 1.2, the string representation omits trailing zeroes. But in Hive 2.3, it is always padded to 18 digits with trailing zeroes if necessary.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: `the string representation`-> `the decimal string representation`

Comment (cloud-fan) [reply docs/sql-migration-guide.md]: TRANSFORM syntax -> TRANSFORM operator

**Review by cloud-fan — APPROVED** [review]

**Comment by Ngone51** [comment]
Jenkins, retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by Ngone51** [comment]
thanks all!!

**Pull request closed without merging**